frontend/public/module/k8s/k8s-actions.js

```diff
@@ -155,21 +152,25 @@ const actions = {
     dispatch({type: types.watchK8sList, id, query});
     REF_COUNTS[id] = 1;
 
-    // Fetch entire list (XHR) then use its resourceVersion to
-    //  start listening on a WS (?resourceVersion=$resourceVersion)
-    //  start the process over when:
-    //   1. the WS closes abnormally
-    //   2. the WS can not establish a connection within $TIMEOUT
+    const incrementallyLoad = async(continueToken = '') => {
+      const response = await k8sList(k8skind, {...query, limit: paginationLimit, continue: continueToken}, true);
```

Comment:
A couple of things:

1. It looks like `incrementallyLoad` will run to completion even if we drop the watch (`stopK8sWatch`).  I didn't look too closely, but we may even continue to start up the rest of the machinery (this could be an existing race condition).  We should probably check `REF_COUNTS[id]` on every iteration and `throw()` and do magic in the catch.  

2. I can imagine `incrementallyLoad`ing never finishing on a large cluster with a sad internet.  The behavior in that case would be we start loading the list, a XHR drops, then we throw away the data and start over.  What do you think about adding a try/catch here to retry any given continueToken N times?

Reply:
I added some basic retry logic (`attempts <= 3`) and a naïve attempt to cancel the incremental `fetch` if a `stopK8sWatch` is dispatched. I'm not sure the best way to `catch` the errors I throw, any suggestions @kans?

Reply:
Errors should fall into the error handler on the `then` - (second function so we don't accidentally catch rendering errors) on [219](https://github.com/alecmerdler/console/blob/18cfbb5f9e1c644830766d95fc0ed783566ca73e/frontend/public/module/k8s/k8s-actions.js#L219).  It needs to check for ¿both? refcounts and the strings you put in the errors or you could subclass Error and check for that for more flexibility/clarity.  

BTW, I don't think throwing strings is cromulent anymore.

Reply:
@kans I'm having trouble getting an intuition around these edge cases; would you be okay with merging what we have here and making follow-up PRs to address those? I plan to improve our load-testing tools (ServiceWorker, etc) once this is merged.

Reply:
I think the worst case behavior would be that we create WS that live forever.  To fix the race, we'd need to check for `REF_COUNTS[id]` in three places (and throw a *special Error*):

1. before [#L160](https://github.com/alecmerdler/console/blob/2fca07b00f1e29b3ad912169c8b948dac81cdc2b/frontend/public/module/k8s/k8s-actions.js#L160)
2. Before we create the WS on [#L180](https://github.com/alecmerdler/console/blob/2fca07b00f1e29b3ad912169c8b948dac81cdc2b/frontend/public/module/k8s/k8s-actions.js#180).  
3. A new function in the [promise.all](https://github.com/alecmerdler/console/blob/2fca07b00f1e29b3ad912169c8b948dac81cdc2b/frontend/public/module/k8s/k8s-actions.js#L176) because `pollAndWatch` is called recursively for error handling.  

We catch all errors on #L213.  If the error is *special error*, I think we just need to exit.

I'd be OK deferring this work to an *immediate* follow up since we are already subject to the same race.